streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -91,14 +91,14 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final Intern
         try {
             context.setRecordContext(entry.recordContext());
             if (flushListener != null) {
-
                 final V oldValue = sendOldValues ? serdes.valueFrom(underlying.get(entry.key())) : null;
+                underlying.put(entry.key(), entry.newValue());
```

Comment:
Is this actually needed? If a downstream processor tries to get this value won't it get it from the cache? i.e., i don't think an evicted entry is removed from the cache until after flush has finished

Reply:
Actually it will be removed, i.e. when ` listener.apply(entries);` is called the entry is no longer in the cache anymore.

What I observed originally is an issue when caching is turned off: note that with that case we still go through this code path (which should be optimized away anyways in the future I think), when you call a `put` on the store, it will immediately trigger flush and hence be processed downstream while it is not put into the underlying and also not in the cache any more.